### PR TITLE
Adds a partition key to the reminders table, to resolve #22

### LINKED ISSUE #22: Reminders needs a partition key to support shared throughput
With the case of wanting to store both the Application Data and the Orleans data within the same CosmosDB instance, where that CosmosDB instance is using the new "shared" throughput facilities, it is a requirement to set partition keys for each collection, according to the output [below].

```
Lifecycle start canceled due to errors at stage 20000
warn: Orleans.Runtime.Scheduler *stg/22/00000016.WorkItemGroup[101215]
      Task [Id=11, Status=RanToCompletion] in WorkGroup [SystemTarget: S10.1.1.18:11111:282256074*stg/22/00000016@S00000016] took elapsed time 0:00:01.1214201 for execution, which is longer than 00:00:00.2000000. Running on thread System.Threading.Thread
dbug: Orleans.Runtime.DeploymentLoadPublisher[0]
      UpdateRuntimeStatistics from S10.1.1.18:11111:282256074
   at Orleans.LifecycleSubject.OnStart(CancellationToken ct)
   at Orleans.Runtime.Scheduler.AsyncClosureWorkItem.Execute()
   at Orleans.Runtime.Silo.StartAsync(CancellationToken cancellationToken)
   at Adapt.Hegemon.TMS.Silo.Program.StartSilo() in D:\src\Adapt\Hegemon-TMS\src\Adapt.Hegemon.TMS.Silo\Program.cs:line 154
   at Adapt.Hegemon.TMS.Silo.Program.Main(String[] args) in D:\src\Adapt\Hegemon-TMS\src\Adapt.Hegemon.TMS.Silo\Program.cs:line 130
--------------------------------------------------------------------------------
Provisioned throughput collection should have a partition key
ActivityId: 44fa21ea-8e0a-43ad-a6e2-a5bf0d909d80, Microsoft.Azure.Documents.Common/2.2.0.0, Linux/9 documentdb-netcore-sdk/1.9.1
   at Microsoft.Azure.Documents.Client.ClientExtensions.ParseResponseAsync(HttpResponseMessage responseMessage, JsonSerializerSettings serializerSettings)
   at Microsoft.Azure.Documents.GatewayStoreModel.<>c__DisplayClass20_0.<<InvokeAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Azure.Documents.BackoffRetryUtility`1.<>c__DisplayClass1_0.<<ExecuteAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Azure.Documents.BackoffRetryUtility`1.ExecuteRetry(Func`1 callbackMethod, Func`3 callShouldRetry, CancellationToken cancellationToken, Action`1 preRetryCallback)
   at Microsoft.Azure.Documents.BackoffRetryUtility`1.ExecuteRetry(Func`1 callbackMethod, Func`3 callShouldRetry, CancellationToken cancellationToken, Action`1 preRetryCallback)
   at Microsoft.Azure.Documents.BackoffRetryUtility`1.ExecuteAsync(Func`1 callbackMethod, IRetryPolicy retryPolicy, CancellationToken cancellationToken, Action`1 preRetryCallback)
   at Microsoft.Azure.Documents.GatewayStoreModel.InvokeAsync(DocumentServiceRequest request, ResourceType resourceType)
   at Microsoft.Azure.Documents.GatewayStoreModel.ProcessMessageAsync(DocumentServiceRequest request)
   at Microsoft.Azure.Documents.Client.DocumentClient.CreateAsync(DocumentServiceRequest request)
   at Microsoft.Azure.Documents.Client.DocumentClient.CreateDocumentCollectionPrivateAsync(String databaseLink, DocumentCollection documentCollection, RequestOptions options)
   at Microsoft.Azure.Documents.BackoffRetryUtility`1.<>c__DisplayClass1_0.<<ExecuteAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Azure.Documents.BackoffRetryUtility`1.ExecuteRetry(Func`1 callbackMethod, Func`3 callShouldRetry, CancellationToken cancellationToken, Action`1 preRetryCallback)
   at Microsoft.Azure.Documents.BackoffRetryUtility`1.ExecuteRetry(Func`1 callbackMethod, Func`3 callShouldRetry, CancellationToken cancellationToken, Action`1 preRetryCallback)
   at Microsoft.Azure.Documents.BackoffRetryUtility`1.ExecuteAsync(Func`1 callbackMethod, IRetryPolicy retryPolicy, CancellationToken cancellationToken, Action`1 preRetryCallback)
   at Microsoft.Azure.Documents.Client.DocumentClient.CreateDocumentCollectionIfNotExistsPrivateAsync(Uri databaseUri, DocumentCollection documentCollection, RequestOptions options)
   at Orleans.Reminders.CosmosDB.CosmosDBReminderTable.TryCreateCosmosDBResources()
   at Orleans.Reminders.CosmosDB.CosmosDBReminderTable.Init()
   at Orleans.OrleansTaskExtentions.WithTimeout(Task taskToComplete, TimeSpan timeout, String exceptionMessage)
   at Orleans.Runtime.ReminderService.LocalReminderService.Start()
   at Orleans.Runtime.Scheduler.AsyncClosureWorkItem.Execute()
   at Orleans.OrleansTaskExtentions.WithTimeout(Task taskToComplete, TimeSpan timeout, String exceptionMessage)
   at Orleans.Runtime.Silo.<OnActiveStart>g__StartReminderService|74_0()
   at Orleans.Runtime.Silo.StartAsyncTaskWithPerfAnalysis(String taskName, Func`1 task, Stopwatch stopWatch)
   at Orleans.Runtime.Silo.OnActiveStart(CancellationToken ct)
   at Orleans.Runtime.SiloLifecycleSubject.MonitoredObserver.OnStart(CancellationToken ct)
   at Orleans.LifecycleSubject.WrapExecution(CancellationToken ct, Func`2 action)
   at Orleans.LifecycleSubject.OnStart(CancellationToken ct)
warn: Orleans.Runtime.Silo[100220]
```

### PR DESCRIPTION
**Pull request by dealproc**

### CODE REVIEW

File: src/Orleans.Clustering.CosmosDB/Orleans.Clustering.CosmosDB.csproj [hunk 1]
```diff
@@ -62,7 +62,7 @@
   </ItemGroup>
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.Azure.DocumentDB.Core" Version="1.9.1" />
+    <PackageReference Include="Microsoft.Azure.DocumentDB.Core" Version="[1.9.1,2.0)" />
```
Comment (galvesribeiro) [inline src/Orleans.Clustering.CosmosDB/Orleans.Clustering.CosmosDB.csproj]: Unrelated to this PR. It should be somewhere else. Please revert it.

File: src/Orleans.Persistence.CosmosDB/Orleans.Persistence.CosmosDB.csproj [hunk 2]
```diff
@@ -50,7 +50,7 @@
   </ItemGroup>
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.Azure.DocumentDB.Core" Version="1.9.1" />
+    <PackageReference Include="Microsoft.Azure.DocumentDB.Core" Version="[1.9.1,2.0)" />
```
Comment (galvesribeiro) [inline src/Orleans.Persistence.CosmosDB/Orleans.Persistence.CosmosDB.csproj]: Unrelated.

File: src/Orleans.Reminders.CosmosDB/CosmosDBReminderTable.cs [hunk 3]
```diff
@@ -278,7 +279,7 @@ private async Task TryCreateCosmosDBResources()
             {
                 Id = this._options.Collection
             };
-
+            remindersCollection.PartitionKey.Paths.Add(PARTITION_KEY);
```
Comment (galvesribeiro) [inline src/Orleans.Reminders.CosmosDB/CosmosDBReminderTable.cs]: Just doing that is not goind to work.

```
"   at Microsoft.Azure.Documents.Client.DocumentClient.AddPartitionKeyInformationAsync(DocumentServiceRequest request, RequestOptions options)\n   at Microsoft.Azure.Documents.Client.DocumentClient.ExecuteStoredProcedurePrivateAsync[TValue](String storedProcedureLink, RequestOptions options, Object[] procedureParams)\n   at Microsoft.Azure.Documents.BackoffRetryUtility`1.<>c__DisplayClass1_0.<<ExecuteAsync>b__0>d.MoveNext()\n--- End of stack trace from previous location where exception was thrown ---\n   at Microsoft.Azure.Documents.BackoffRetryUtility`1.ExecuteRetry(Func`1 callbackMethod, Func`3 callShouldRetry, CancellationToken cancellationToken, Action`1 preRetryCallback)\n   at Microsoft.Azure.Documents.BackoffRetryUtility`1.ExecuteRetry(Func`1 callbackMethod, Func`3 callShouldRetry, CancellationToken cancellationToken, Action`1 preRetryCallback)\n   at Microsoft.Azure.Documents.BackoffRetryUtility`1.ExecuteAsync(Func`1 callbackMethod, IRetryPolicy retryPolicy, CancellationToken cancellationToken, Action`1 preRetryCallback)\n   at Orleans.Reminders.CosmosDB.CosmosDBReminderTable.ExecuteWithRetries[TResult](Func`1 clientFunc) in /Users/guto/dev/repos/OrleansContrib/Orleans.CosmosDB/src/Orleans.Reminders.CosmosDB/CosmosDBReminderTable.cs:line 234\n   at Orleans.Reminders.CosmosDB.CosmosDBReminderTable.ReadRows(UInt32 begin, UInt32 end) in /Users/guto/dev/repos/OrleansContrib/Orleans.CosmosDB/src/Orleans.Reminders.CosmosDB/CosmosDBReminderTable.cs:line 150"
```

All the tests fail. Did you ran them?

Also, can you elaborate on why `GrainId` should be a partition key? Please check the other providers to come out with something that will not affect performance.

The CI isn't enabled on this repo yet, so please make sure you `dotnet test` before submit the PR.

File: src/Orleans.Reminders.CosmosDB/Orleans.Reminders.CosmosDB.csproj [hunk 4]
```diff
@@ -58,7 +58,7 @@
   </ItemGroup>
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.Azure.DocumentDB.Core" Version="1.9.1" />
+    <PackageReference Include="Microsoft.Azure.DocumentDB.Core" Version="[1.9.1,2.0)" />
```
Comment (galvesribeiro) [inline src/Orleans.Reminders.CosmosDB/Orleans.Reminders.CosmosDB.csproj]: Unrelated.

File: src/Orleans.Clustering.CosmosDB/Orleans.Clustering.CosmosDB.csproj (hunk 1 above)
Comment (dealproc) [reply src/Orleans.Clustering.CosmosDB/Orleans.Clustering.CosmosDB.csproj]: Will reply once for all three.  I do agree with this comment, and the changes are made in all three locations.  I will make a separate branch to contribute the nuget change(s) required, or please do feel free do so as you see fit.

Comment (dealproc) [reply src/Orleans.Clustering.CosmosDB/Orleans.Clustering.CosmosDB.csproj]: Seeing what I am getting at, I will revoke this PR if you are willing to make the necessary changes.  I've been trying to sort out the necessary changes, but they're not quickly clicking.  There are a few edge cases that seem to need additional thought and/or deeper knowledge of how the storage mechanism works, which I do not have.

Comment (galvesribeiro) [reply src/Orleans.Clustering.CosmosDB/Orleans.Clustering.CosmosDB.csproj]: If you fix the tests failing I'm ok merging it. I'm just saying the package upgrade here should be a different PR and as Orleans 2.2 is about to be released, I would hold on with that change because I'll make the Orleans upgrade as soon as it is out.

**Comment by galvesribeiro** [comment]
Thanks for the PR but it looks like just adding the PK on the table definition will not sort your issues (if any).

Once I got the packages updated to Orleans 2.2 I'll check the _new_ collection RU allocation modes and see if it does affect us anyhow.

I'm closing this PR for now.
